SYCL/ESIMD/api/functional/value_conv.hpp

```diff
@@ -0,0 +1,117 @@
+//===-- value_conv.hpp - This file provides common functions generate values for
+//      testing. ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides functions that let obtain converted reference data for
+/// test according to current underlying types.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "value.hpp"
+
+namespace esimd_test::api::functional {
+
+// Utility class to retrieve specific values for tests depending on the source
+// and destination data types. May be used to retrieve converted reference data.
+// All provided methods are save to use and protected from UB when call
+// static_cast<int>(unsigned int).
+template <typename SrcT, typename DstT> struct value_conv {
+  static SrcT lowest() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
+      if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+        return static_cast<SrcT>(value<DstT>::lowest());
+      } else {
+        return value<SrcT>::lowest();
+      }
+    } else {
+      return std::max(value<SrcT>::lowest(),
+                      static_cast<SrcT>(value<DstT>::lowest()));
+    }
+  }
+
+  static SrcT max() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
+      if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+        return static_cast<SrcT>(value<DstT>::max());
+      } else {
+        return value<SrcT>::max();
+      }
+    } else {
+      return std::min(value<SrcT>::max(),
+                      static_cast<SrcT>(value<DstT>::max()));
+    }
+  }
```

Comment:
I suggest do it in separate PR because we need more time to avoid UB cases
I leave the TODO comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/5959fb1f2f4dd5aca794998f3a625824bfc42107#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R52-R53)